zkfacade/src/main/java/com/yahoo/vespa/curator/Lock.java

```diff
@@ -20,54 +20,50 @@
  */
 public class Lock implements Mutex {
 
-    private final ReentrantLock lock;
     private final InterProcessLock mutex;
     private final String lockPath;
 
     public Lock(String lockPath, Curator curator) {
+        this(lockPath, curator.createMutex(lockPath));
+    }
+
+    /** Public for testing only */
+    public Lock(String lockPath, InterProcessLock mutex) {
         this.lockPath = lockPath;
-        this.lock = new ReentrantLock(true);
-        mutex = curator.createMutex(lockPath);
+        this.mutex = mutex;
     }
 
     /** Take the lock with the given timeout. This may be called multiple times from the same thread - each matched by a close */
     public void acquire(Duration timeout) throws UncheckedTimeoutException {
+        ThreadLockInfo threadLockInfo = ThreadLockInfo.getCurrentThreadLockInfo();
+        threadLockInfo.invokingAcquire(lockPath, timeout);
+
+        final boolean acquired;
         try {
-            if ( ! mutex.acquire(timeout.toMillis(), TimeUnit.MILLISECONDS))
-                throw new UncheckedTimeoutException("Timed out after waiting " + timeout +
-                                                    " to acquire lock '" + lockPath + "'");
-            if ( ! lock.tryLock()) { // Should be available to only this thread, while holding the above mutex.
-                release();
-                throw new IllegalStateException("InterProcessMutex acquired, but guarded lock held by someone else, for lock '" + lockPath + "'");
-            }
-        }
-        catch (UncheckedTimeoutException | IllegalStateException e) {
-            throw e;
-        }
-        catch (Exception e) {
+            acquired = mutex.acquire(timeout.toMillis(), TimeUnit.MILLISECONDS);
+        } catch (Exception e) {
+            threadLockInfo.acquireFailed(lockPath);
             throw new RuntimeException("Exception acquiring lock '" + lockPath + "'", e);
         }
-    }
 
-    @Override
-    public void close() {
-        try {
-            lock.unlock();
-        }
-        finally {
-            release();
+        if (!acquired) {
+            threadLockInfo.acquireTimedOut(lockPath);
+            throw new UncheckedTimeoutException("Timed out after waiting " + timeout +
+                    " to acquire lock '" + lockPath + "'");
         }
+        threadLockInfo.lockAcquired(lockPath);
     }
 
-    private void release() {
+    @Override
+    public void close() {
+        ThreadLockInfo.getCurrentThreadLockInfo().lockReleased(lockPath);
```

Comment:
Shouldn't this happen after the `mutex.release()` below?

Reply:
To get any Instant.now() to include the time in release?

Reply:
No, currently if `mutex.release()` fails  you will still have tagged it as released in `ThreadLockInfo`.

Reply:
I could count an error, but the lock must be assumed to be released whether an exception is thrown or not

Reply:
Added count